### PR TITLE
feat: refactor error to be more clear and maintainable

### PR DESCRIPTION
**Pull request by PureWhiteWu**
1. All the thrift exceptions should be defined together, so this pr moves `ApplicationError` from volo to pilota.
2. The previous `XError` such as `ApplicationError`, `ProtocolError`, `TransportError` correspond to the exception concept in thrift rpc protocol, this pr renames those types to `XException`.
3. The `Error` type does not have a clear semantic, this pr renames it to `ThriftException` which is clearer in meaning.
4. The `write_x` methods in `rw_ext.rs` never returns error, so this PR removes the `Result` type from function sig.
5. Previously the `TransportException` only comes from `std::io::Error` and its kind is a subset of `std::io::ErrorKind`, and separate a new `TransportExceptionKind` from `std::io::ErrorKind` does not make sense. So this pr deletes the `TransportExceptionKind` and uses the `std::io::Error` directly.
6. Previously there's a standalone type called `DecodeError` which is a mixture of some variants of `ThriftException`, and the reason it exists is to add more information about the context of the error. But the implementation has some issues, such as we cannot easily get the root kind of error. The only thing we need to do is to prepend a message to the error, so this pr adds a `prepend_msg` function to `ThriftException` and uses it as the codec error, which is far clearer and more maintainable.

### CODE REVIEW

**Comment by GuangmingLuo** [comment]
请根据 PR 模板和规范，填写 Motivation 和 Solution，否则这个模板和规范存在没有意义。

**Comment by PureWhiteWu** [comment]
> 请根据 PR 模板和规范，填写 Motivation 和 Solution，否则这个模板和规范存在没有意义。

已完成。

File: pilota-build/test_data/thrift/default_value.rs.bak
Comment (wfly1998) [inline pilota-build/test_data/thrift/default_value.rs.bak:1]: Forget to delete?

File: pilota-build/test_data/thrift/normal.rs.bak
Comment (wfly1998) [inline pilota-build/test_data/thrift/normal.rs.bak:1]: Ditto

File: pilota/src/thrift/error/protocol.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            ProtocolExceptionKind::Unknown => "protocol error",
+            ProtocolExceptionKind::InvalidData => "bad data",
+            ProtocolExceptionKind::NegativeSize => "negative message size",
+            ProtocolExceptionKind::SizeLimit => "message too long",
+            ProtocolExceptionKind::BadVersion => "invalid thrift version",
+            ProtocolExceptionKind::NotImplemented => "not implemented",
+            ProtocolExceptionKind::DepthLimit => "maximum skip depth reached",
+        };
+
+        write!(f, "{}, {}", error_text, self.message)
```
Comment (wfly1998) [inline pilota/src/thrift/error/protocol.rs:80]: `{}: {}` would be better?

File: pilota/src/thrift/error/protocol.rs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+pub type ProtocolErrorKind = ProtocolExceptionKind;
+
+/// Runtime library error categories.
+///
+/// This list may grow, and it is not recommended to match against it.
+#[non_exhaustive]
+#[derive(Clone, Copy, Eq, Debug, PartialEq)]
+pub enum ProtocolExceptionKind {
+    /// Catch-all runtime-library error.
+    Unknown,
```
Comment (wfly1998) [inline pilota/src/thrift/error/protocol.rs:99]: `Other` would be better?

File: pilota/src/thrift/error/application.rs [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+            ApplicationExceptionKind::MISSING_RESULT => "missing method result",
+            ApplicationExceptionKind::INTERNAL_ERROR => "remote service threw exception",
+            ApplicationExceptionKind::PROTOCOL_ERROR => "protocol error",
+            ApplicationExceptionKind::INVALID_TRANSFORM => "invalid transform",
+            ApplicationExceptionKind::INVALID_PROTOCOL => "invalid protocol requested",
+            ApplicationExceptionKind::UNSUPPORTED_CLIENT_TYPE => "unsupported protocol client",
+            _ => "other error",
+        };
+
+        write!(f, "{}, msg: {}", error_text, self.message)
```
Comment (wfly1998) [inline pilota/src/thrift/error/application.rs:104]: Redundant `msg` in format

File: pilota/src/thrift/error/application.rs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+///
+/// This list may grow, and it is not recommended to match against it.
+#[non_exhaustive]
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+#[repr(transparent)]
+pub struct ApplicationExceptionKind(i32);
+
+impl ApplicationExceptionKind {
+    /// Catch-all application error.
+    pub const UNKNOWN: Self = Self(0);
```
Comment (wfly1998) [inline pilota/src/thrift/error/application.rs:247]: Try `Other`?

File: pilota/src/thrift/error/application.rs [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+    /// Service reply is missing required fields.
+    pub const MISSING_RESULT: Self = Self(5);
+    /// User server handler error.
+    pub const INTERNAL_ERROR: Self = Self(6);
+    /// Thrift protocol error. When possible use `Error::ProtocolError` with a
+    /// specific `ProtocolExceptionKind` instead.
+    pub const PROTOCOL_ERROR: Self = Self(7);
+    /// *Unknown*. Included only for compatibility with existing Thrift
+    /// implementations.
+    pub const INVALID_TRANSFORM: Self = Self(8); // ??
```
Comment (wfly1998) [inline pilota/src/thrift/error/application.rs:267]: ??

File: pilota/src/thrift/error/transport.rs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    /// Prepend a message to the existing error message.
+    ///
+    /// That means, the new message will be: `message` + `old_message`.
+    pub fn prepend_msg(&mut self, message: &str) {
+        let mut s = String::with_capacity(self.message.len() + message.len());
+        s.push_str(message);
+        s.push_str(self.message.as_str());
+        self.message = s.into();
+    }
+}
```
Comment (wfly1998) [inline pilota/src/thrift/error/transport.rs:72]: These functions are the same for all `Exception`s, maybe we can define a macro for them?

File: pilota/src/thrift/error/protocol.rs (hunk 2 above)
Comment (PureWhiteWu) [reply pilota/src/thrift/error/protocol.rs:83]: I think we should be consistent with other languages here.

**Review by yukiiiteru — APPROVED** [review]

**Review by Millione — APPROVED** [review]

**Pull request merged**